include/deal.II/base/aligned_vector.h

```diff
@@ -1036,6 +1124,405 @@ AlignedVector<T>::fill(const T &value)
+  // Now that all processes know the address of the space that is visible to
+  // everyone, we need to figure out whether it is properly aligned and if not,
+  // find the next aligned address.
+  //
+  // std::align does that, but it also modifies its last two arguments. The
+  // documentation of that function at
+  // https://en.cppreference.com/w/cpp/memory/align is not entirely clear, but I
+  // *think* that the following should do given that we do not use base_ptr and
+  // available_space any further after the call to std::align.
+  std::size_t available_space       = alloc_size;
+  void *      base_ptr_backup       = base_ptr;
+  T *         aligned_shmem_pointer = static_cast<T *>(
+    std::align(align_by, array_size * sizeof(T), base_ptr, available_space));
+  Assert(aligned_shmem_pointer != nullptr, ExcInternalError());
+
+  // There is one step to guard against. It is *conceivable* that the base_ptr
+  // we have previously obtained from MPI_Win_shared_query is mapped so
+  // awkwardly into the different MPI processes' memory spaces that it is
+  // aligned in one memory space, but not another. In that case, different
+  // processes would align base_ptr differently, and adjust available_space
+  // differently. We can check that by making sure that the max (or min) over
+  // all processes is equal to every process's value. If that's not the case,
+  // then the whole idea of aligning above is wrong and we need to rethink what
+  // it means to align data in a shared memory space.
+  //
+  // One might be tempted to think that this is not how MPI implementations
+  // actually arrange things. Alas, when developing this functionality in 2021,
+  // this is really how at least OpenMPI ends up doing things. (This is with an
+  // OpenMPI implementation of MPI 3.1, so it does not support the flag we set
+  // in the MPI_Info structure above when allocating the memory window.) Indeed,
+  // when running this code on three processes, one ends up with base_ptr values
+  // of
+  //     base_ptr=0x7f0842f02108
+  //     base_ptr=0x7fc0a47881d0
+  //     base_ptr=0x7f64872db108
+  // which, most annoyingly, are aligned to 8 and 16 byte boundaries -- so there
+  // is no common offset std::align could find that leads to a 64-byte
+  // aligned memory address in all three memory spaces. That's a tremendous
+  // nuisance and there is really nothing we can do about this other than just
+  // fall back on the (unaligned) base_ptr in that case.
+  if (Utilities::MPI::min(available_space, shmem_group_communicator) !=
+      Utilities::MPI::max(available_space, shmem_group_communicator))
+    aligned_shmem_pointer = static_cast<T *>(base_ptr_backup);
+
+
+  // **** Step 6 ****
+  // If this is the shmem root process, we need to copy the data into the
+  // shared memory space.
+  if (is_shmem_root)
+    {
+      if (std::is_trivial<T>::value == true)
+        std::memcpy(aligned_shmem_pointer, elements.get(), sizeof(T) * size());
```

Comment:
Can't we broadcast into this address instead of using a separate buffer first and then do the memcpy?

Reply:
I asked the same earlier. I think we should let this go and optimize in a follow up. I want this feature in 9 .3 and optimizations can wait.